src/engine.ml

```diff
@@ -1892,24 +1892,27 @@ let resolve_connect_client config ts s ev =
     if retry_exceeded retry' then
       Error (`Msg "maximum connection retries exceeded")
     else
-      Ok
-        (match v with
+      let state, action =
+        match v with
         | `Domain (name, ip_version), _, _ ->
             (Resolving (idx', ts, retry'), `Resolve (name, ip_version))
         | `Ip ip, port, dp ->
-            (Connecting (idx', ts, retry'), `Connect (ip, port, dp)))
+            (Connecting (idx', ts, retry'), `Connect (ip, port, dp))
+      in
+      Ok ({ t with linger = Cstruct.empty; state = client state }, action)
   in
   match (s, ev) with
   | Resolving (idx, _, retry), `Resolved ip ->
       (* TODO enforce ipv4/ipv6 *)
       let endp = match remote idx with _, port, dp -> (ip, port, dp) in
-      Ok (Connecting (idx, ts, retry), Some (`Connect endp))
+      let t = { t with state = client (Connecting (idx, ts, retry)) } in
```

Comment:
here we don't need to clean up the linger? I understand that the invariant we try to establish is that when we're in "connecting" state, there is no linger, is this correct?

Reply:
Yes, we want an invariant that when the state is `Connecting` (or `Resolving`) linger is empty.

Reply:
let's write that down somewhere then :)